### PR TITLE
routing: persistent mission control

### PR DESCRIPTION
**Pull request by joostjager**
This PR adds persistence to mission control by storing the raw payment results and reprocessing them on startup.

### CODE REVIEW

**Comment by joostjager** [comment]
Reminder: extract error source index as an int, in case a node appears in a route twice.

**Comment by joostjager** [comment]
note to self: handle private channels

**Comment by joostjager** [comment]
Ran a small performance test to check startup time impact of rebuilding mission control state. On my machine reconstructing from 1000 historical payment results took 15 ms.

**Comment by Roasbeef** [comment]
No longer WIP and ready to review?

**Comment by joostjager** [comment]
wip label removed.

File: routing/payment_session.go [hunk 1]
```diff
@@ -16,27 +16,6 @@ type PaymentSession interface {
 	// specified HTLC payment to the target node.
 	RequestRoute(payment *LightningPayment,
 		height uint32, finalCltvDelta uint16) (*route.Route, error)
-
```
Comment (Roasbeef) [inline routing/payment_session.go]: 👍 clean separation of concerns

File: routing/missioncontrol.go [hunk 2]
```diff
@@ -369,3 +369,215 @@ func (m *MissionControl) GetHistorySnapshot() *MissionControlSnapshot {
...
+	var (
+		failureVertex route.Vertex
+	)
+	if failureSourceIdx > 0 {
+		failureVertex = rt.Hops[failureSourceIdx-1].PubKeyBytes
+	} else {
+		failureVertex = rt.SourcePubKey
+	}
+
+	// Always determine chan id ourselves, because a channel
```
Comment (Roasbeef) [inline routing/missioncontrol.go]: All the line wrapping in this method looks like it can be updated to fill out each line more.

File: routing/missioncontrol.go [hunk 3]
```diff
@@ -373,7 +373,7 @@ func (m *MissionControl) GetHistorySnapshot() *MissionControlSnapshot {
 // reportPaymentFail reports a failed payment to mission control as input for
 // future probability estimates. It returns a bool indicating whether this error
 // is a final error and no further payment attempts need to be made.
-func (m *MissionControl) reportPaymentFail(rt *route.Route,
+func (m *MissionControl) reportPaymentFail(paymentID uint64, rt *route.Route,
```
Comment (Roasbeef) [inline routing/missioncontrol.go]: Why does it need the payment ID? Atm this is a lower level identifier that shouldn't really impact the operation of mission control.

File: routing/missioncontrol.go [hunk 4]
```diff
@@ -126,6 +126,17 @@ type MissionControlChannelSnapshot struct {
 	SuccessProb float64
 }
 
+// paymentResult is the information that becomes available when a payment
```
Comment (Roasbeef) [inline routing/missioncontrol.go]: Commit title+body are pretty bare, just mentions payment results just then makes independent changes to other methods.

File: routing/missioncontrol.go [hunk 5]
```diff
@@ -18,6 +19,9 @@ const (
 	// minSecondChanceInterval is the minimum time required between
 	// second-chance failures.
 	minSecondChanceInterval = time.Minute
+
+	// DefaultMaxMcHistory is the default maximum history size.
```
Comment (Roasbeef) [inline routing/missioncontrol.go]: Does this include failed attempts as well?

File: lnrpc/routerrpc/config.go [hunk 6]
```diff
@@ -25,4 +25,8 @@ type RoutingConfig struct {
 	// AprioriHopProbability is the assumed success probability of a hop in
 	// a route when no other information is available.
 	AprioriHopProbability float64
+
+	// MaxMcHistory defines the maximum number of payment results that
+	// are held on disk by mission control.
+	MaxMcHistory int `long:"maxmchistory" description:"the maximum number of payment results that are held on disk by mission control"`
```
Comment (Roasbeef) [inline lnrpc/routerrpc/config.go]: Also this struct has many fields that are duplicated below.

File: routing/missioncontrol.go [hunk 7]
```diff
@@ -138,29 +148,67 @@ type paymentResult struct {
...
+	if err := mc.init(); err != nil {
+		return nil, err
+	}
+
+	return mc, nil
+}
+
+// init initializes mission control with historical data.
+func (m *MissionControl) init() error {
+	log.Debugf("Mission control state reconstruction started")
```
Comment (Roasbeef) [inline routing/missioncontrol.go]: Perhaps log timing information on the `Tracef` log level along with the number of entries loaded? I think we can even just do this asynchronously, blocking other prospective callers to mission control if their requests arrive before we've finished loading all the prior history.

File: routing/missioncontrol_store.go [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+		if err := tx.DeleteBucket(resultsKey); err != nil {
+			return err
+		}
+
+		_, err := tx.CreateBucket(resultsKey)
+		return err
+	})
+}
+
+// fetch returns all results currently stored in the database.
```
Comment (Roasbeef) [inline routing/missioncontrol_store.go]: `fetchAll`? 

`retrieveAll`? 

`loadAll`?

File: routing/missioncontrol_store.go [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+func (b *missionControlStore) AddResult(rp *paymentResult) error {
+	return b.db.Update(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(resultsKey)
+
+		// Prune oldest entries.
+		if b.maxRecords > 0 {
+			count := bucket.Stats().KeyN
+			for count >= b.maxRecords {
+				cursor := bucket.Cursor()
+				cursor.First()
```
Comment (Roasbeef) [inline routing/missioncontrol_store.go]: Errors not checked.

File: routing/missioncontrol_store.go [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+			&b,
+			uint64(rp.timeFwd.UnixNano()),
+			uint64(rp.timeReply.UnixNano()),
+			rp.success, dbFailureSourceIdx,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Write route.
```
Comment (Roasbeef) [inline routing/missioncontrol_store.go]: [No comments are better than these bare ones which don't add any additional context or commentary.](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#43-code-documentation-and-commenting)

File: routing/missioncontrol_store.go [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+		if err := channeldb.SerializeRoute(&b, *rp.route); err != nil {
+			return err
+		}
+
+		// Write failure.
```
Comment (Roasbeef) [inline routing/missioncontrol_store.go]: Possibly use `WriteVarBytes` here instead? The failure will just be blank upon read if nothing was specified.

File: routing/missioncontrol_store.go [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+		}
+
+		// Put into results bucket.
+		return bucket.Put(getKey(rp), b.Bytes())
+	})
+}
+
+// idToKey returns a byte slice representing a unique key for this payment
+// attempt.
+func getKey(rp *paymentResult) []byte {
```
Comment (Roasbeef) [inline routing/missioncontrol_store.go]: Mismatched function name and `godoc` comment. Also `getKey` is pretty vague.

**Comment by joostjager** [comment]
> IMO the scope of this PR doesn't match the title nor the body. Reading the title I'd expect the bulk of the diff to be focused on persistence and resurrection logic. Instead, this PR contains a number of changes, many of which can stand as isolated PRs to be evaluated independently.

Yes, there is definitely a lot of preparation before the actual persistence is added. I needed to have all mission control logic (second chance logic, failure attribution) in one place to be able to record and playback payment results. 

I can take the last commit `routing: persist mission control data` off if you want?

File: lnrpc/routerrpc/config.go [hunk 13]
```diff
@@ -0,0 +1,28 @@
+package routerrpc
+
+import (
+	"time"
+
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// RoutingConfig contains the configurable parameters that control routing.
+type RoutingConfig struct {
```
[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/routerrpc/config.go]: It would be a circular dependency.

[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/routerrpc/config.go]: Ah no, that is not the reason. There is no circular dependency.

The reason that it is here, is because it is a helper struct that only exists to deal with the conditional compilation and different variants of the sub server config struct. Without conditional compilation, we could have just taken those values from `routerrpc.Config`. 

I have expressed my opinion on conditional compilation multiple times before. In short, I advocate for dropping conditional compilation and clearly marking experimental rpc calls as experimental in the docs. 

This code and discussion is again something that for me counts towards unnecessary complications without upside.

File: lnrpc/routerrpc/config_active.go [hunk 14]
```diff
@@ -72,10 +72,9 @@ func DefaultConfig() *Config {
 	}
 }
 
-// GetMissionControlConfig returns the mission control config based on this sub
-// server config.
-func GetMissionControlConfig(cfg *Config) *routing.MissionControlConfig {
-	return &routing.MissionControlConfig{
+// GetRoutingConfig returns the routing config based on this sub server config.
```
[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/routerrpc/config_active.go]: Do you mean the command line parameters? They are here because they also will be here eventually when there is no more main rpc.

File: routing/payment_session_source.go [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+	SelfNode *channeldb.LightningNode
+
+	// MissionControl is a shared memory of sorts that executions of payment
+	// path finding use in order to remember which vertexes/edges were
+	// pruned from prior attempts. During payment execution, errors sent by
+	// nodes are mapped into a vertex or edge to be pruned. Each run will
+	// then take into account this set of pruned vertexes/edges to reduce
+	// route failure and pass on graph information gained to the next
+	// execution.
+	MissionControl missionControlInterface
```
[the remark this answers is not in the document]
Comment (joostjager) [reply routing/payment_session_source.go]: Is there a reason to make it public? There is no use of it outside of `routing`.

File: routing/router.go [hunk 16]
```diff
@@ -171,6 +171,18 @@ type PaymentSessionSource interface {
 	NewPaymentSessionEmpty() PaymentSession
 }
 
+// missionControlInterface is an interface that exposes failure reporting and
```
[the remark this answers is not in the document]
Comment (joostjager) [reply routing/router.go]: Yes, I know. I didn't succeed in coming up with different names for mission control the implementation and mission control the interface. What would you suggest?

File: routing/router_test.go [hunk 17]
```diff
@@ -263,7 +263,7 @@ func TestSendPaymentRouteFailureFallback(t *testing.T) {
 	var payHash [32]byte
 	paymentAmt := lnwire.NewMSatFromSatoshis(1000)
 	payment := LightningPayment{
-		Target:      ctx.aliases["luoji"],
+		Target:      ctx.aliases["sophon"],
```
[the remark this answers is not in the document]
Comment (joostjager) [reply routing/router_test.go]: I am aware of these cases. But I still think we shouldn't rely on mission control to handle them. In my model of how this should work, such an attempt will fail and by the time we start another path finding round, the local channel data is updated and that channel won't be used anymore.

File: routing/missioncontrol.go [hunk 18]
```diff
@@ -16,6 +14,10 @@ const (
 	// half-life duration defines after how much time a penalized node or
 	// channel is back at 50% probability.
 	DefaultPenaltyHalfLife = time.Hour
+
+	// minSecondChanceInterval is the minimum time required between
+	// second-chance failures.
```
[the remark this answers is not in the document]
Comment (joostjager) [reply routing/missioncontrol.go]: Added

File: routing/missioncontrol.go [hunk 19]
```diff
@@ -209,6 +216,39 @@ func (m *MissionControl) getEdgeProbabilityForNode(nodeHistory *nodeHistory,
...
+// at providing a channel update for its channel with toNode.
+func (m *MissionControl) requestSecondChance(timestamp time.Time,
+	fromNode, toNode route.Vertex) bool {
+
+	// Look up previous second chance time.
+	pair := DirectedNodePair{
+		From: fromNode,
+		To:   toNode,
+	}
+	lastSecondChance := m.lastSecondChance[pair]
```
[the remark this answers is not in the document]
Comment (joostjager) [reply routing/missioncontrol.go]: fixed

File: routing/missioncontrol.go [hunk 20]
```diff
@@ -209,6 +216,39 @@ func (m *MissionControl) getEdgeProbabilityForNode(nodeHistory *nodeHistory,
...
+
+		log.Debugf("Second chance granted for %v->%v", fromNode, toNode)
+
+		return true
+	}
+
+	// Otherwise penalize the channel, because we don't allow channel
+	// updates that are that frequent. This is to prevent nodes from keeping
+	// us busy by continuously sending new channel updates.
+	log.Debugf("Second chance denied for %v->%v", fromNode, toNode)
```
[the remark this answers is not in the document]
Comment (joostjager) [reply routing/missioncontrol.go]: Added

File: routing/missioncontrol.go (hunk 2 above)
Comment (joostjager) [reply routing/missioncontrol.go]: Fixed

File: routing/missioncontrol.go (hunk 3 above)
Comment (joostjager) [reply routing/missioncontrol.go]: It is to prevent the same payment result from being recorded multiple times because of `lnd` restarts. It makes the call idempotent.

File: routing/missioncontrol.go (hunk 4 above)
Comment (joostjager) [reply routing/missioncontrol.go]: Commit split in two

File: lnrpc/routerrpc/config.go (hunk 6 above)
Comment (joostjager) [reply lnrpc/routerrpc/config.go]: This is again a consequence of the conditional compilation. Suggestions to remove duplication are welcome.

File: routing/missioncontrol.go (hunk 5 above)
Comment (joostjager) [reply routing/missioncontrol.go]: In this PR we only store failed attempts.

File: routing/missioncontrol.go (hunk 7 above)
Comment (joostjager) [reply routing/missioncontrol.go]: Log added. Kept it on debug level as that felt more in line with other log statements.

Async loading is definitely a good idea, but I wanted to do that outside of this PR.

File: routing/missioncontrol_store.go (hunk 8 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: `fetchAll`

File: routing/missioncontrol_store.go (hunk 9 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: Check added for `Delete()`

File: routing/missioncontrol_store.go (hunk 10 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: I see these as 'chapters'. For `SerializeRoute` in isolation I wouldn't have added it, but leaving it out in this context makes it less readable imo.

File: routing/missioncontrol_store.go (hunk 11 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: Converted to `WriteVarBytes`. I also discovered that I was serializing data that included length and padding. Created a separate commit that defines slimmer encode/decode functions in `lnwire`.

File: routing/missioncontrol_store.go (hunk 12 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: Fixed and renamed

**Comment by joostjager** [comment]
@roasbeef ptal

**Comment by joostjager** [comment]
Another reason why it may be good to take the persistence commit off for the first merge is to test the other changes first and see if that somehow triggers a change in the db structure.

File: routing/router.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply routing/router.go]: `RouteMemory`? lol 

or `PathfindingMemory` 

`RoutingCommand` 

`PaymentHistory`

File: lnrpc/routerrpc/config.go (hunk 6 above)
Comment (Roasbeef) [reply lnrpc/routerrpc/config.go]: Create a struct with the shared fields, and embed it into both of them.

File: routing/missioncontrol_store.go [hunk 21]
```diff
@@ -0,0 +1,235 @@
...
+			return err
+		}
+
+		_, err := tx.CreateBucket(resultsKey)
+		return err
+	})
+}
+
+// fetch returns all results currently stored in the database.
+func (b *missionControlStore) fetchAll() ([]*paymentResult, error) {
```
Comment (Roasbeef) [inline routing/missioncontrol_store.go]: Method name out of sync with the `godoc` comment.

File: routing/missioncontrol_store.go (hunk 10 above)
Comment (Roasbeef) [reply routing/missioncontrol_store.go]: Don't think this one in particular adds anything but an extra line in the diff.

File: routing/payment_session_source.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply routing/payment_session_source.go]: Exported in #3298

File: routing/router.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply routing/router.go]: `MissionController` it is now.

File: routing/missioncontrol_store.go (hunk 21 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: fixed

File: routing/missioncontrol_store.go (hunk 10 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: I still like it the way it is, but it is unimportant. Merged comment with one above.

File: lnrpc/routerrpc/config.go (hunk 6 above)
Comment (joostjager) [reply lnrpc/routerrpc/config.go]: Shared struct created.

File: routing/router.go [hunk 22]
```diff
@@ -1873,38 +1873,40 @@ func (r *ChannelRouter) tryApplyChannelUpdate(rt *route.Route,
...
+			)
+			if err != nil {
+				return true, channeldb.FailureReasonError
+			}
 		}
+
+		log.Tracef("node=%v reported failure when sending htlc",
+			*failureSourceIdx)
+	} else {
+		log.Tracef("unreadable failure when sending htlc")
```
Comment (cfromknecht) [inline routing/router.go]: nit: log mess should be capitalized

File: routing/router.go [hunk 23]
```diff
@@ -1873,38 +1873,40 @@ func (r *ChannelRouter) tryApplyChannelUpdate(rt *route.Route,
...
+		if failureMessage != nil {
+			err := r.tryApplyChannelUpdate(
+				rt, *failureSourceIdx, failureMessage,
+			)
+			if err != nil {
+				return true, channeldb.FailureReasonError
+			}
 		}
+
+		log.Tracef("node=%v reported failure when sending htlc",
```
Comment (cfromknecht) [inline routing/router.go]: capitalization

File: lnrpc/routerrpc/config.go [hunk 24]
```diff
@@ -3,26 +3,25 @@ package routerrpc
...
+	AprioriHopProbability float64 `long:"apriorihopprob" description:"Assumed success probability of a hop in a route when no other information is available."`
+
+	// PenaltyHalfLife defines after how much time a penalized node or
+	// channel is back at 50% probability.
+	PenaltyHalfLife time.Duration `long:"penaltyhalflife" description:"Defines the duration after which a penalized node or channel is back at 50% probability"`
+
+	// AttemptCost is the virtual cost in path finding weight units of
+	// executing a payment attempt that fails. It is used to trade off
+	// potentially better routes against their probability of succeeding.
+	AttemptCost btcutil.Amount `long:"attemptcost" description:"The (virtual) cost in sats of a failed payment attempt"`
```
Comment (cfromknecht) [inline lnrpc/routerrpc/config.go]: what's the rationale for changing this to sats, when we later convert it back to msats? this will make anyone's current config values be off by a factor of 1000 after updating

File: routing/missioncontrol_store.go [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+}
+
+// AddResult adds a new result to the db.
+func (b *missionControlStore) AddResult(rp *paymentResult) error {
+	return b.db.Update(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(resultsKey)
+
+		// Prune oldest entries.
+		if b.maxRecords > 0 {
+			count := bucket.Stats().KeyN
```
Comment (cfromknecht) [inline routing/missioncontrol_store.go]: we've had issues in the past using `Stats()`, perhaps instead we can count the number of entries in the db on startup and track the current count in memory?

File: routing/missioncontrol_store.go [hunk 26]
```diff
@@ -0,0 +1,235 @@
...
+	)
+
+	err := channeldb.ReadElements(
+		r, &timeFwd, &timeReply, &result.success, &dbFailureSourceIdx,
+	)
+	if err != nil {
+		return nil, err
+	}
+	result.timeFwd = time.Unix(0, int64(timeFwd)).UTC()
+	result.timeReply = time.Unix(0, int64(timeReply)).UTC()
```
Comment (cfromknecht) [inline routing/missioncontrol_store.go]: do we use this `UTC()` conversion anywhere else in the codebase?

File: routing/missioncontrol_store.go [hunk 27]
```diff
@@ -0,0 +1,235 @@
...
+		// Write failure. If there is no failure message, write an empty
+		// byte slice.
+		var failureBytes bytes.Buffer
+		if rp.failure != nil {
+			err := lnwire.EncodeFailureMessage(&failureBytes, rp.failure, 0)
+			if err != nil {
+				return err
+			}
+		}
+		err = wire.WriteVarBytes(&b, 0, failureBytes.Bytes())
```
Comment (cfromknecht) [inline routing/missioncontrol_store.go]: definitely not a blocker, but could be nice to have an equivalent `serializeResult` method so that's it's easier to examine to encoding and decoding logic side-by-side

File: routing/missioncontrol_store.go [hunk 28]
```diff
@@ -0,0 +1,235 @@
...
+	// stored.
+	resultsKey = []byte("missioncontrol-results")
+
+	// Big endian is the preferred byte order, due to cursor scans over
+	// integer keys iterating in order.
+	byteOrder = binary.BigEndian
+
+	// unknownFailureSourceIdx is the database encoding of an unknown error
+	// source.
+	unknownFailureSourceIdx = -1
```
Comment (cfromknecht) [inline routing/missioncontrol_store.go]: where is this used?

File: routing/missioncontrol.go [hunk 29]
```diff
@@ -1,6 +1,7 @@
 package routing
 
 import (
+	"github.com/lightningnetwork/lnd/channeldb"
```
Comment (halseth) [inline routing/missioncontrol.go]: import order.

File: routing/router.go [hunk 30]
```diff
@@ -1909,195 +1907,9 @@ func (r *ChannelRouter) processSendError(rt *route.Route, sendErr error) (
...
-			channel: failedEdge.channel,
-		}, 0)
-		return false, 0
-
-	// Any other failure or an empty failure will get the node pruned.
-	default:
-		r.cfg.MissionControl.ReportVertexFailure(failureVertex)
-		return false, 0
-	}
+	return r.cfg.MissionControl.ReportPaymentFail(
```
Comment (halseth) [inline routing/router.go]: Nice 👍

File: routing/missioncontrol.go [hunk 31]
```diff
@@ -393,26 +393,30 @@ func (m *MissionControl) GetHistorySnapshot() *MissionControlSnapshot {
 // future probability estimates. It returns a bool indicating whether this error
 // is a final error and no further payment attempts need to be made.
 func (m *MissionControl) ReportPaymentFail(rt *route.Route,
-	failureSourceIdx int, failure lnwire.FailureMessage) (bool,
+	failureSourceIdx *int, failure lnwire.FailureMessage) (bool,
```
Comment (halseth) [inline routing/missioncontrol.go]: the pointer argument needs godoc

File: routing/missioncontrol.go [hunk 32]
```diff
@@ -393,26 +393,30 @@ func (m *MissionControl) GetHistorySnapshot() *MissionControlSnapshot {
...
 	channeldb.FailureReason) {
 
-	var (
-		failureVertex route.Vertex
-	)
+	var failureSourceIdxInt int
+	if failureSourceIdx == nil {
```
Comment (halseth) [inline routing/missioncontrol.go]: this needs explanation/comments

File: routing/router.go [hunk 33]
```diff
@@ -1873,38 +1873,40 @@ func (r *ChannelRouter) tryApplyChannelUpdate(rt *route.Route,
...
+		failureMessage   lnwire.FailureMessage
+		failureSourceIdx *int
+	)
 
-	// If an internal, non-forwarding error occurred, we can stop trying.
-	fErr, ok := sendErr.(*htlcswitch.ForwardingError)
-	if !ok {
-		return true, channeldb.FailureReasonError
-	}
+	if sendErr != htlcswitch.ErrUnreadableFailureMessage {
```
Comment (halseth) [inline routing/router.go]: would this be easier to follow using a switch statement?

File: routing/router.go [hunk 34]
```diff
@@ -178,7 +178,7 @@ type MissionController interface {
 	// input for future probability estimates. It returns a bool indicating
 	// whether this error is a final error and no further payment attempts
 	// need to be made.
-	ReportPaymentFail(rt *route.Route,
+	ReportPaymentFail(paymentID uint64, rt *route.Route,
```
Comment (halseth) [inline routing/router.go]: what's the purpose of this change?

File: routing/missioncontrol.go [hunk 35]
```diff
@@ -307,32 +307,30 @@ func (m *MissionControl) reportEdgePolicyFailure(failedEdge edge) {
...
 	history := m.createHistoryIfNotExists(failedEdge.from)
-	history.lastFail = &now
+	history.lastFail = &timestamp
 }
 
 // reportEdgeFailure reports a channel level failure.
 //
 // TODO(roasbeef): also add value attempted to send and capacity of channel
-func (m *MissionControl) reportEdgeFailure(failedEdge edge,
+func (m *MissionControl) reportEdgeFailure(timestamp time.Time, failedEdge edge,
```
Comment (halseth) [inline routing/missioncontrol.go]: maybe we could add timestamp to the `edge` struct?

File: lnrpc/routerrpc/config.go (hunk 24 above)
Comment (joostjager) [reply lnrpc/routerrpc/config.go]: It was already in sats. The diff may be a bit confusing, because the command line options struct and the internal config struct are unified in this pr.

File: routing/missioncontrol_store.go (hunk 26 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: I did a quick find and it seems we don't. The reason I do this is to get UTC timezones for log messages that contain these times.

File: routing/router.go (hunk 34 above)
Comment (joostjager) [reply routing/router.go]: To make reporting to mission control idempotent. Without this id, it wouldn't be able to recognize when it is offered the same result again after restart.

**Comment by joostjager** [comment]
> Since persisting mission control data is a big change, I would suggesting breaking this up into two or more PRs:
> 
> 1. Simple changes (exporting Decode/Serialization methods, embedding configs)
> 2. Changing params passed to mission control, and interfaces
> 3. Writing data to disk

We could move the last commit that finally adds the persistence to its own pr. But not sure if it makes sense at this stage of the review process.

File: routing/missioncontrol.go (hunk 31 above)
Comment (joostjager) [reply routing/missioncontrol.go]: more godoc added explaining the ptr

File: routing/missioncontrol.go (hunk 32 above)
Comment (joostjager) [reply routing/missioncontrol.go]: comments got lost when I moved this code. readded

File: routing/router.go (hunk 33 above)
Comment (joostjager) [reply routing/router.go]: not sure exactly how to convert to a switch, but i took another look at this part and simplified it in another way.

File: routing/missioncontrol.go (hunk 35 above)
Comment (joostjager) [reply routing/missioncontrol.go]: I prefer to keep them separately. The `edge` is instantiated in `getFailedEdge` and this function has no idea of time.

File: routing/missioncontrol_store.go (hunk 28 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: it should have been used instead of constants in the code. Fixed.

File: routing/missioncontrol_store.go (hunk 25 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: Do you have more details on what issues those were? We can work around it, but it would be nice to have some confirmation that `Stats().KeyN` is indeed unreliable.

File: routing/missioncontrol_store.go (hunk 27 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: added

File: routing/missioncontrol_store.go (hunk 26 above)
Comment (cfromknecht) [reply routing/missioncontrol_store.go]: UTC is is the default https://play.golang.org/p/e_-nuERpM66

File: routing/missioncontrol_store.go (hunk 25 above)
Comment (cfromknecht) [reply routing/missioncontrol_store.go]: it caused a panic in one of our prior migrations:
 * issue: https://github.com/lightningnetwork/lnd/issues/1874#issuecomment-420022023
 * fix: https://github.com/lightningnetwork/lnd/pull/1898

the root cause was never determined, so we've been avoiding `Stats()` like the plague

File: lnrpc/routerrpc/config.go (hunk 24 above)
Comment (cfromknecht) [reply lnrpc/routerrpc/config.go]: ah i see, thanks for explaining

File: routing/missioncontrol_store.go (hunk 26 above)
Comment (wpaulino) [reply routing/missioncontrol_store.go]: I think that's just because the server being used for the playground is on UTC. Running this locally results in my timezone being shown instead.

Comment (cfromknecht) [reply routing/missioncontrol_store.go]: hmm, that's obnoxious.

but still the bigger question is why wouldn't we print this in the user's local timezone? nothing else in our logs are UTCified

Comment (joostjager) [reply routing/missioncontrol_store.go]: I assumed we want UTC in order to be able to easily compare/merge logs from different machines without needing to think about physical location. But if we don't do it anywhere else in the code base, we may not want to do it here either.

File: routing/missioncontrol_store.go (hunk 25 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: Ah, I see. So there is no other way to get a count than iterating through all items with a cursor?

Comment (joostjager) [reply routing/missioncontrol_store.go]: Ah, I see. So there is no other way to get a count than iterating through all items with a cursor?

Comment (joostjager) [reply routing/missioncontrol_store.go]: The issue that you mention, could it be that it was caused by a database that was already corrupt?

Comment (cfromknecht) [reply routing/missioncontrol_store.go]: > Ah, I see. So there is no other way to get a count than iterating through all items with a cursor?

yeah that's the simplest way

> The issue that you mention, could it be that it was caused by a database that was already corrupt?

we saw it happen on perfectly good dbs. once the fix landed, people were able to resume w/o issue afaiaa

File: routing/missioncontrol_store.go (hunk 26 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: UTC removed

File: routing/missioncontrol_store.go (hunk 25 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: `Stats()` call removed. Small complication because store needs to be idempotent, but solved that with an additional bucket get.

How did you actually know these were perfectly good dbs?

File: routing/router_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply routing/router_test.go]: Although this change is part of a commit that is now removed from the pr, I found a case where we are unfortunately still forced to penalize local channels.

When the error source is unknown, it may or may not be our peer. In that case we can only penalize all channels along the route including our own.

File: routing/missioncontrol_store.go [hunk 36]
```diff
@@ -0,0 +1,271 @@
...
+	})
+}
+
+// fetchAll returns all results currently stored in the database.
+func (b *missionControlStore) fetchAll() ([]*paymentResult, error) {
+	var results []*paymentResult
+
+	err := b.db.View(func(tx *bbolt.Tx) error {
+		resultBucket := tx.Bucket(resultsKey)
+		results = make([]*paymentResult, resultBucket.Stats().KeyN)
```
Comment (cfromknecht) [inline routing/missioncontrol_store.go]: still calling `Stats().KeyN` here. i think it's fine if we just don't allocate it up front and append. if this isn't used outside of testing then it shouldn't be a performance issue

File: routing/missioncontrol_store.go [hunk 37]
```diff
@@ -0,0 +1,271 @@
...
+func (b *missionControlStore) AddResult(rp *paymentResult) error {
+	return b.db.Update(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(resultsKey)
+
+		// Prune oldest entries.
+		if b.maxRecords > 0 {
+			for b.numRecords >= b.maxRecords {
+				cursor := bucket.Cursor()
+				cursor.First()
+				if err := cursor.Delete(); err != nil {
```
Comment (cfromknecht) [inline routing/missioncontrol_store.go]: in the past i think we observed a performance issue in calling `First()` then `Delete()` repetitively, since it actually results in a quadratic operation. fortunately i think we only expect this to be called once here, but something to keep in mind in the future. @Roasbeef was this not one of the root causes in that migration performance issue?

File: routing/missioncontrol_store.go (hunk 36 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: Yes my mistake. I had fixed that already on a different computer, but hadn't pushed.

File: routing/missioncontrol_store.go (hunk 37 above)
Comment (joostjager) [reply routing/missioncontrol_store.go]: Ok interesting. Yes, it is expected to be called only once.

Comment (Roasbeef) [reply routing/missioncontrol_store.go]: IIRC, in that case, we would have repeated calls to `Delete()`, then `First()`, as `Delete` would move the cursor one record after the deleted key. We ended up resolving it by using a two pass approach: https://github.com/lightningnetwork/lnd/commit/974ec02d249e6a77712e24233692e4e08c8bf31b#diff-99f3013528876f55a0e46ff7014e8a19
